driver/ixsystems/common.py

```diff
@@ -435,6 +438,48 @@ def _create_volume_from_snapshot(self, name, snapshot_name,
         except Exception as e:
             raise FreeNASApiError('Unexpected error', e)
 
+    def _promote_volume(self, volume_name):
+        """Promote a volume"""
+        request_urn = ('%s/id/%s/promote') % (
+            FreeNASServer.REST_API_VOLUME,
+            urllib.parse.quote_plus(
+                self.configuration.ixsystems_dataset_path + '/' + volume_name))
+        LOG.debug('_promote_volume urn : %s', request_urn)
+        try:
+            ret = self.handle.invoke_command(FreeNASServer.CREATE_COMMAND,
+                                             request_urn, None)
+            if ret['status'] != FreeNASServer.STATUS_OK:
+                msg = ('Error while promoting volume: %s' % ret['response'])
+                raise FreeNASApiError('Unexpected error', msg)
+        except Exception as e:
+            raise FreeNASApiError('Unexpected error', e)
+
+    def _is_service_project(self, project_id):
+        # Use keystone api to check project_id is service project
+        # Return True if it is service project, otherwise return False
+        grp = cfg.OptGroup('keystone_authtoken')
+        ops = [cfg.StrOpt('auth_url'), cfg.StrOpt('username'), cfg.StrOpt('password'),
+               cfg.StrOpt('project_name'), cfg.StrOpt('user_domain_name'), cfg.StrOpt('project_domain_name')]
+        CONF.register_group(grp)
+        CONF.register_opts(ops, group=grp)
+        auth = v3.Password(auth_url=CONF.keystone_authtoken.auth_url,
+                           username=CONF.keystone_authtoken.username,
+                           password=CONF.keystone_authtoken.password,
+                           project_id=project_id,
+                           user_domain_name=CONF.keystone_authtoken.user_domain_name,
+                           project_domain_name=CONF.keystone_authtoken.project_domain_name)
+        sess = session.Session(auth=auth)
+        keystone = client.Client(session=sess)
+        try:
+            project = keystone.projects.get(project_id)
+            if project.name == CONF.keystone_authtoken.project_name:
+                return True
+        except Exception:
+            # Invalid project id will cause exeception from keystone client,
+            # in this case it is allowed and normal, hence do nothing
+            pass
+        return False
+
```

Comment:
To understand a project_id is actually service project id is by calling keystone api retrieve project object.